### PR TITLE
BIP155: AddrV2

### PR DESCRIPTION
**Pull request by jrawsthorne**
Thought I'd just open this to see what people's thoughts are on how to do this. 

- I guess we don't want to implement parsing for tor, i2p and cjdns addresses so do we just want to have them all as unknown or do we want to do what I've done here and just store the raw bytes
- Not sure about names
- Do we want to store time separately like it is for addrv1

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
I think raw bytes is right. Anyone using those protocols (eg my dns seed will probably start using v3 onions as soon as core does) will almost certainly just dump the bytes into XXXX.onion and pass it to some other code so more parsing is just harder.

File: src/network/address.rs [hunk 1]
```diff
@@ -110,12 +110,186 @@ impl fmt::Debug for Address {
...
 }
 
+/// Address received from BIP155 addrv2 message
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct AddrV2Address {
+    /// Time that this node was last seen as connected to the network
+    pub time: u32,
+    /// Service bits
+    pub services: ServiceFlags,
+    /// IP address
```
Comment (ariard) [inline src/network/address.rs:120]: Not necessary an IP address, I think it should say "Network ID + Network address. The interpretation depends on networkID". Just copy the BIP.

File: src/network/address.rs [hunk 2]
```diff
@@ -110,12 +110,186 @@ impl fmt::Debug for Address {
     }
 }
 
+/// Address received from BIP155 addrv2 message
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct AddrV2Address {
```
Comment (ariard) [inline src/network/address.rs:115]: W.r.t naming, maybe `AddrV2Msg` for the message and just `AddrV2` thus avoiding the reference to IP.

File: src/network/address.rs [hunk 3]
```diff
@@ -191,5 +365,113 @@ mod test {
...
+
+        // Invalid IPv4, with bogus length.
+        assert!(deserialize::<AddrV2IP>(&Vec::from_hex("010501020304").unwrap()).is_err());
+
+        // Invalid IPv4, with extreme length.
+        assert!(deserialize::<AddrV2IP>(&Vec::from_hex("01fd010201020304").unwrap()).is_err());
+
+        // Valid IPv6.
+        let ip: AddrV2IP = deserialize(&Vec::from_hex("02100102030405060708090a0b0c0d0e0f10").unwrap()).unwrap();
+        assert_eq!(ip, AddrV2IP::IPV6(Ipv6Addr::from_str("102:304:506:708:90a:b0c:d0e:f10").unwrap()));
```
Comment (ariard) [inline src/network/address.rs:414]: Missing "Valid IPv6. contains embedded "internal" ?

**Review by ariard — COMMENTED** [review]
Quick survey around test correspondence from Core.

File: src/network/address.rs (hunk 3 above)
Comment (jrawsthorne) [reply src/network/address.rs]: It looked like internal addresses were an implementation detail of bitcoin core

File: src/network/address.rs [hunk 4]
```diff
@@ -110,12 +110,169 @@ impl fmt::Debug for Address {
...
+    pub services: ServiceFlags,
+    /// Network ID + Network Address
+    pub addr: AddrV2,
+    /// Network port, 0 if not applicable
+    pub port: u16
+}
+
+/// Supported networks for use in BIP155 addrv2 message
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum AddrV2 {
```
Comment (stevenroose) [inline src/network/address.rs:128]: nit: I think Rust convention is to not capitalize abbreviations in type/variant names. So `Ipv6`, `Cjdns` etc. I don't feel strongly about this, though.

**Review by stevenroose — APPROVED** [review]
Only left a few nits, but I think it's good enough like this!

File: src/network/address.rs [hunk 5]
```diff
@@ -110,12 +110,169 @@ impl fmt::Debug for Address {
...
+    I2P([u8; 32]),
+    /// CJDNS
+    CJDNS(Ipv6Addr),
+    /// Unknown
+    Unknown(u8, Vec<u8>),
+}
+
+impl Encodable for AddrV2 {
+    fn consensus_encode<W: io::Write>(&self, mut e: W) -> Result<usize, encode::Error> {
+        macro_rules! encode_addr {
```
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/network/address.rs:147]: It seems that @sgeisler is at least leaning towards avoiding macros in favor of functions: https://github.com/rust-bitcoin/rust-bitcoin/pull/478#discussion_r497772695

**Comment by jrawsthorne** [comment]
@stevenroose Addressed all your comments. I agree with replacing the macro with a function, less room for error.

**Review by stevenroose — APPROVED** [review]
Awesome, LGTM!

**Review by apoelstra — APPROVED** [review]
ack 363e1106adddb495670660ffdda1ebe30582c096

I'm not thrilled with the `vec` allocation when we receive an unknown message type, but I think it's probably the most reasonable thing.

**Comment by apoelstra** [comment]
ah needs rebase.

**Comment by apoelstra** [comment]
Lol finally kicked Travis into working

**Review by apoelstra — APPROVED** [review]
ack 921f64699c25b478da5ecdffbf2bb9cb386d1fc9

**Review by stevenroose — APPROVED** [review]
utACK

**Pull request merged**